### ISSUE TITLE
Broken authorization key? AttributeError: 'NoneType' object has no attribute 'key'

### ISSUE THREAD

**Issue by arrix** [LonamiWebs/Telethon#907]
Telethon 1.0.2
I have a long running telethon program that listens to messages from over 100 megagroups using `client.add_event_handler`. It receives messages as expected but can only last at most a few hours. From the log below we can see that the `Broken authorization key` error occurred 8 hours after the last message. There should be a ton of messages in between. When the error happens, the session authorization will usually be expired or (more likely) corrupted `telethon.errors.rpcerrorlist.AuthKeyUnregisteredError: The key is not registered in the system`.

```
2018-07-24 16:17:52 Your verify code : 7758C9 is activated !
2018-07-25 00:18:06,860 WARNING mtprotosender.py#455 Broken authorization key?: Can't decode packet (b'S\xfe\xff\xff')
2018-07-25 00:18:06,861 ERROR base_events.py#1268 Task exception was never retrieved
future: <Task finished coro=<MTProtoSender._reconnect() done, defined at /home/aaa/venvs/myenv/lib/python3.6/site-packages/telethon/network/mtprotosender.py:286> exception=AttributeError("'NoneType' object has no attribute 'key'",)>
Traceback (most recent call last):
  File "/home/aaa/venvs/myenv/lib/python3.6/site-packages/telethon/network/mtprotosender.py", line 294, in _reconnect
    await self._send_loop_handle
  File "/home/aaa/venvs/myenv/lib/python3.6/site-packages/telethon/network/mtprotosender.py", line 379, in _send_loop
    body = self.state.pack_message(message)
  File "/home/aaa/venvs/myenv/lib/python3.6/site-packages/telethon/network/mtprotostate.py", line 88, in pack_message
    self.auth_key.key[88:88 + 32] + data + padding).digest()
AttributeError: 'NoneType' object has no attribute 'key'
```

I backup the session time before starting the program and restore from it when this happens. The backup session is still authorized.

Another variation of error often seen is

```
2018-07-24 23:59:35,063 INFO mtprotosender.py#641 Message 6581813862618524076 not resent due to bad msg
2018-07-24 23:59:35,063 INFO mtprotosender.py#641 Message 6581813862633175376 not resent due to bad msg
2018-07-24 23:59:35,067 INFO mtprotosender.py#641 Message 6581813862662022116 not resent due to bad msg
2018-07-24 23:59:35,067 INFO mtprotosender.py#641 Message 6581813862676605704 not resent due to bad msg
2018-07-24 23:59:35,070 WARNING mtprotosender.py#455 Broken authorization key?: Can't decode packet (b'S\xfe\xff\xff')
2018-07-24 23:59:35,070 ERROR base_events.py#1268 Task exception was never retrieved
future: <Task finished coro=<MTProtoSender._reconnect() done, defined at /home/aaa/venvs/myenv/lib/python3.6/site-packages/telethon/network/mtprotosender.py:286> exception=AttributeError("'NoneType' object has no attribute 'key'",)>
Traceback (most recent call last):
  File "/home/aaa/venvs/myenv/lib/python3.6/site-packages/telethon/network/mtprotosender.py", line 294, in _reconnect
    await self._send_loop_handle
  File "/home/aaa/venvs/myenv/lib/python3.6/site-packages/telethon/network/mtprotosender.py", line 379, in _send_loop
    body = self.state.pack_message(message)
  File "/home/aaa/venvs/myenv/lib/python3.6/site-packages/telethon/network/mtprotostate.py", line 88, in pack_message
    self.auth_key.key[88:88 + 32] + data + padding).digest()
AttributeError: 'NoneType' object has no attribute 'key'
```

And System clock is wrong often happens before it
```
2018-07-25 00:17:53,634 INFO mtprotosender.py#611 System clock is wrong, set time offset to -21s
2018-07-25 00:17:53,638 INFO mtprotosender.py#611 System clock is wrong, set time offset to -21s
2018-07-25 00:17:53,642 INFO mtprotosender.py#611 System clock is wrong, set time offset to -21s
... hundreds of lines ...
2018-07-25 00:17:56,173 INFO mtprotosender.py#611 System clock is wrong, set time offset to -23s
2018-07-25 00:17:56,176 INFO mtprotosender.py#611 System clock is wrong, set time offset to -23s
2018-07-25 00:17:56,180 INFO mtprotosender.py#611 System clock is wrong, set time offset to -23s
2018-07-25 00:17:56,184 INFO mtprotosender.py#611 System clock is wrong, set time offset to -23s
2018-07-25 00:17:56,187 INFO mtprotosender.py#611 System clock is wrong, set time offset to -23s
```

**Comment by Lonami** [comment]
> Telethon 1.0.2

Upgrade.

> I backup the session time before starting the program and restore from it when this happens

And it works (**please** answer this)? That's good to know. Maybe it should simply log a warning instead erasing the key. Not sure why these occur from time to time. I have never encountered such issue.

> And System clock is wrong often happens before it

Pretty weird the message is spammed in the log so fast (4ms delay between bad message notifications). The library should be fixing its own time offset, but for some reason that's not working for you. What does the `DEBUG` log say?

**Comment by Lonami** [comment]
Maybe I should update the message ID for all pending messages if these occur, huh. This worked a bit differently pre-1.0, so maybe that's it.

**Comment by arrix** [comment]
Yes, the backup session always works. I've tried a dozen times.

**Comment by arrix** [comment]
Happened again in 1.1.1 master version. 
I'm sure I have `logging.basicConfig(level=logging.DEBUG)` but not DEBUG log seen.

```
2018-07-26 11:06:16,665 INFO mtprotosender.py#649 Message 6582356798548890744 not resent due to bad msg
2018-07-26 11:06:16,667 INFO mtprotosender.py#649 Message 6582356798552531872 not resent due to bad msg
2018-07-26 11:06:16,668 INFO mtprotosender.py#649 Message 6582356798557180084 not resent due to bad msg
2018-07-26 11:06:16,669 INFO mtprotosender.py#649 Message 6582356798561215080 not resent due to bad msg
2018-07-26 11:06:16,670 INFO mtprotosender.py#649 Message 6582356798565131820 not resent due to bad msg
2018-07-26 11:06:16,671 INFO mtprotosender.py#649 Message 6582356798569197332 not resent due to bad msg
2018-07-26 11:06:16,672 INFO mtprotosender.py#649 Message 6582356798573092136 not resent due to bad msg
2018-07-26 11:06:16,673 INFO mtprotosender.py#649 Message 6582356798577118552 not resent due to bad msg
2018-07-26 11:06:16,674 INFO mtprotosender.py#649 Message 6582356798581896460 not resent due to bad msg
2018-07-26 11:06:16,675 WARNING mtprotosender.py#457 Broken authorization key?: Can't decode packet (b'S\xfe\xff\xff')
2018-07-26 11:06:16,676 ERROR base_events.py#1268 Task exception was never retrieved
future: <Task finished coro=<MTProtoSender._reconnect() done, defined at /home/aaa/venvs/myenv/lib/python3.6/site-packages/telethon/network/mtprotosender.py:286> exception=AttributeError("'NoneType' object has no attribute 'key'",)>
Traceback (most recent call last):
  File "/home/aaa/venvs/myenv/lib/python3.6/site-packages/telethon/network/mtprotosender.py", line 294, in _reconnect
    await self._send_loop_handle
  File "/home/aaa/venvs/myenv/lib/python3.6/site-packages/telethon/network/mtprotosender.py", line 379, in _send_loop
    body = self.state.pack_message(message)
  File "/home/aaa/venvs/myenv/lib/python3.6/site-packages/telethon/network/mtprotostate.py", line 89, in pack_message
    self.auth_key.key[88:88 + 32] + data + padding).digest()
AttributeError: 'NoneType' object has no attribute 'key'

```

**Comment by arrix** [comment]
@Lonami Is there anyway for user code to catch or get notified of such errors?

I have `client.add_event_handler(my_event_handler)` and `await client.disconnected`, currently the program doesn't exit.

**Comment by arrix** [comment]
@Lonami 

> Maybe it should simply log a warning instead erasing the key.

I'd like to try to implement this. Could you point to me to some relevant code paths?
Thank you!

**Comment by Lonami** [comment]
That's cool, thank you for the effort. Look around here:
https://github.com/LonamiWebs/Telethon/blob/f2c8663266dddd5d08024fead5b01c9d8e50bba6/telethon/network/mtprotosender.py#L445-L460

**Comment by christosR** [comment]
@arrix are you behind a proxy?

**Comment by Lonami** [comment]
Bump @arrix. Any progress?

**Comment by arrix** [comment]
@Lonami No, actually it never happened again since my last comment.
Now the only error stopping my long running program is
ConnectionError
```
ConnectionError
...
telethon.errors.rpcbaseerrors.BadMessageError: Message too old, and it cannot be verified whether the server has received a message with this msg_id or not.
```

But that's another issue.

From my limited understanding, the auth_key can be created again if erased and is computed according to data received from server. I think the code is correct and the problem is gone because telegram fixed something on the server side.

**Comment by Lonami** [comment]
May be worth trying `master`.

**Comment by nqzhang** [comment]
i have same issue
`Task exception was never retrieved
future: <Task finished coro=<MTProtoSender._reconnect() done, defined at C:\Users\zhangnengqin\AppData\Local\Programs\Python\Python36-32\lib\site-packages\telethon\network\mtprotosender.py:288> exception=AttributeError("'NoneType' object has no attribute 'key'",)>
Traceback (most recent call last):
  File "C:\Users\zhangnengqin\AppData\Local\Programs\Python\Python36-32\lib\site-packages\telethon\network\mtprotosender.py", line 296, in _reconnect
    await self._send_loop_handle
  File "C:\Users\zhangnengqin\AppData\Local\Programs\Python\Python36-32\lib\site-packages\telethon\network\mtprotosender.py", line 383, in _send_loop
    body = self.state.pack_message(message)
  File "C:\Users\zhangnengqin\AppData\Local\Programs\Python\Python36-32\lib\site-packages\telethon\network\mtprotostate.py", line 90, in pack_message
    self.auth_key.key[88:88 + 32] + data + padding).digest()
AttributeError: 'NoneType' object has no attribute 'key'`

**Comment by Lonami** [comment]
@nqzhang did you try installing master.

**Comment by nqzhang** [comment]
> @nqzhang did you try installing master.

My installed version is Telethon==1.3

I'll install from the master branch and see
